Xamarin.Forms.Core.UnitTests/DataTemplateSelectorTests.cs

```diff
@@ -90,4 +90,69 @@ public void NestingThrowsException ()
 			Assert.Throws<NotSupportedException> (() => dts.SelectTemplate ((byte)0, null));
 		}
 	}
+
+	[TestFixture]
+	public class DataTemplateRecycleTests : BaseTestFixture
+	{
+		[TearDown]
+		public override void TearDown()
+		{
+			base.TearDown();
+			Device.PlatformServices = null;
+		}
+
+		[SetUp]
+		public override void Setup()
+		{
+			base.Setup();
+			Device.PlatformServices = new MockPlatformServices();
+		}
+
+		class TestDataTemplateSelector : DataTemplateSelector
+		{
+			readonly DataTemplate declarativeTemplate;
+			readonly DataTemplate proceduralTemplate;
+
+			public TestDataTemplateSelector ()
+			{
+				declarativeTemplate = new DataTemplate(typeof(ViewCell));
+				proceduralTemplate = new DataTemplate(() => new EntryCell());
+			}
+
+			protected override DataTemplate OnSelectTemplate(object item, BindableObject container)
+			{
+				Counter++;
+
+				if (item is string)
+					return declarativeTemplate;
+
+				return proceduralTemplate;
+			}
+
+			public int Counter = 0;
+		}
+
+		[Test]
+		public void ListViewSupport ()
+		{
+			var listView = new ListView(ListViewCachingStrategy.RecycleElementAndDataTemplate);
+			listView.ItemsSource = new object[] { "foo", "bar", 0 };
+
+			Assert.That(listView.CachingStrategy == 
+				ListViewCachingStrategy.RecycleElementAndDataTemplate);
+
+			var selector = new TestDataTemplateSelector();
+			listView.ItemTemplate = selector;
+			Assert.That(selector.Counter == 0);
+
+			Assert.IsInstanceOf<ViewCell>(listView.TemplatedItems[0]);
+			Assert.That(selector.Counter == 1);
+
+			Assert.IsInstanceOf<ViewCell>(listView.TemplatedItems[1]);
+			Assert.That(selector.Counter == 1);
```

Comment:
Resolving the item for the second TemplatedItem hits the DataTemplate cache and so the counter remains at 1.